### ISSUE TITLE
Slow StringTermsAggregatorFromFilters

### ISSUE THREAD

**Issue by sennav** [elastic/elasticsearch#76104]
**Elasticsearch version** (`bin/elasticsearch --version`): 7.13.4 (image from dockerhub)

**Plugins installed**: 
```
[
    discovery-ec2,
    analysis-stempel,
    elasticsearch-analysis-ik,
]
```

**JVM version** (`java -version`):
```
OpenJDK Runtime Environment AdoptOpenJDK (build 16+36)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 16+36, mixed mode, sharing)
```

**OS version** (`uname -a` if on a Unix-like system):
```
Linux 6e9a5be24414 5.10.25-linuxkit #1 SMP Tue Mar 23 09:27:39 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

**Description of the problem including expected versus actual behavior**:

We upgraded our dev environment from 7.6.1 to 7.13.4 today and we noticed our queries became very slow, reading the documentation and troubleshooting our setup we discovered that the aggregation performance was severely degraded especially when the aggregation uses `StringTermsAggregatorFromFilters`. These are screenshots comparing our dev environment against the production environment:

![Screenshot 2021-08-04 at 16 26 08](https://user-images.githubusercontent.com/4136900/128205749-dde68e9f-3329-4411-a273-ec31dea72759.png)

![Screenshot 2021-08-04 at 16 26 42](https://user-images.githubusercontent.com/4136900/128205786-9bc68129-a690-4d37-9804-c5099b4410d7.png)

The dev index has only ~8k documents while the production index has ~30k documents

The relevant part of the aggregation query looks like this:

```
    "Author": {
      "terms": {
        "field": "Author.keyword",
        "order": {
          "_key": "asc"
        },
        "size": 1000
      },
      "aggs": {
        "unique": {
          "cardinality": {
            "field": "id.keyword"
          }
        }
      }
    },
```

Here's a detailed breakdown of the slow aggregation:
![Screenshot 2021-08-04 at 17 22 59](https://user-images.githubusercontent.com/4136900/128208264-8ab729af-a545-4553-a533-ce75a19411c1.png)

I don't know much about the internals of elastic but it could be that `build_leaf_collector` is slow because we aggregate over ids (which might not be ideal) but our indexes are not that big and this seems like a real performance hit compared to version 7.6.1 so that's why I decided to post this here.

I don't have the slowlogs yet because our setup doesn't seem to be forwarding them but I can try to get them and post them here if they are relevant. Let me know if you need any more info and I'm happy to help with anything else you need.

**Comment by not-napoleon** [comment]
Hi @sennav ,

Unfortunately this is a know bug.  We released what we thought was an optimization, but it turns out in some situations it makes things slower.  We have a [fix for this in 7.14](https://github.com/elastic/elasticsearch/pull/74260) so if upgrading is an option for you, that should fix the issue.  If you can't upgrade right away, there's a setting to completely disable the optimization, which should get back to the performance you had before.  You can set `search.aggs.rewrite_to_filter_by_filter: false` to disable it.  See the [PR here](https://github.com/elastic/elasticsearch/pull/73670).

Hope that helps!

**Comment by nik9000** [comment]
> in some situations it makes things slower.

This was my bug and that's a little too kind - in *most* situations it was slower. Upgrading is probably the best solution here - we'll only run the optimization when we're sure it is faster. The setting will disable the "optimization" in all cases, including cases where it would have been faster. It was the "quick escape hatch" we wrote as soon as we found the bug.
